lsp-mode.el

```diff
@@ -2087,24 +2087,73 @@ The `:global' workspace is global one.")
+              'local-map (let ((map (make-sparse-keymap)))
+                           (define-key map [header-line mouse-1]
+                             (lambda ()
+                               (interactive)
+                               (lsp--headerline-breadcrumb-go-to-symbol symbol)))
+                           (define-key map [header-line mouse-2]
+                             (lambda ()
+                               (interactive)
+                               (lsp--headerline-breadcrumb-narrow-to-symbol symbol)))
+                           map)))
+
+(defun lsp--headerline-build-symbols-string (symbols-hierarchy)
+  "Build the symbols breadcrumb from SYMBOLS-HIERARCHY."
   (seq-reduce (lambda (last-symbol-name symbol-to-append)
-                (let ((symbol2-name (if (lsp:document-symbol-deprecated? symbol-to-append)
-                                        (propertize (lsp:document-symbol-name symbol-to-append)
-                                                    'font-lock-face 'lsp-headerline-breadcrumb-deprecated-face)
-                                      (propertize (lsp:document-symbol-name symbol-to-append)
-                                                  'font-lock-face lsp-headerline-breadcrumb-face)))
-                      (symbol2-icon (lsp--headerline-breadcrumb-symbol-icon symbol-to-append))
-                      (arrow-icon (lsp--headerline-breadcrumb-arrow-icon)))
+                (let* ((symbol2-name (if (lsp:document-symbol-deprecated? symbol-to-append)
+                                         (propertize (lsp:document-symbol-name symbol-to-append)
+                                                     'font-lock-face 'lsp-headerline-breadcrumb-deprecated-face)
+                                       (propertize (lsp:document-symbol-name symbol-to-append)
+                                                   'font-lock-face lsp-headerline-breadcrumb-face)))
+                       (symbol2-icon (lsp--headerline-breadcrumb-symbol-icon symbol-to-append))
+                       (arrow-icon (lsp--headerline-breadcrumb-arrow-icon))
+                       (full-symbol-2 (if symbol2-icon
+                                          (concat symbol2-icon symbol2-name)
+                                        symbol2-name)))
                   (format "%s %s %s"
                           last-symbol-name
                           arrow-icon
-                          (if symbol2-icon
-                              (concat symbol2-icon symbol2-name)
-                            symbol2-name))))
+                          (lsp--headerline-with-action symbol-to-append full-symbol-2))))
               symbols-hierarchy ""))
 
+(defun lsp--headerline-dirs-until-root (root-path path)
+  "Find recursively the folders until the project ROOT-PATH.
+PATH is the current folder to be checked."
+  (let ((cur-path (list (f-base path))))
+    (if (f-same? root-path (f-parent path))
```

Comment:
`lsp-f-same?`, unless you want to follow the file symlink

Reply:
Btw, will this function to be called frequently? Following the symlink is quite slow. So you may want to rewrite `f-parent` so it's not using `f-same?` too.

Reply:
Good catch, implemented `lsp-f-parent`